### PR TITLE
RATIS-1539. Refactor GrpcTlsConfig.

### PR DESCRIPTION
**Pull request by szetszwo**
See https://issues.apache.org/jira/browse/RATIS-1539

### CODE REVIEW

**Review by adoroszlai — COMMENTED** [review]
Thanks @szetszwo for working on this.

I have built Ratis with this change and tried Ozone's `ozonesecure-ha` acceptance test.

```
cd hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/ozonesecure-ha
./test.sh
```

Datanodes fail to start due to the following problem:

```
datanode1_1  | 2022-02-28 11:14:23,432 [main] WARN server.RaftServer: FAILED "new RaftServerProxy", attempt #1/5: java.lang.NullPointerException: key required for servers, sleep 500ms and then retry.
datanode1_1  | java.lang.NullPointerException: key required for servers
datanode1_1  |  at org.apache.ratis.thirdparty.io.netty.util.internal.ObjectUtil.checkNotNull(ObjectUtil.java:33)
datanode1_1  |  at org.apache.ratis.thirdparty.io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:434)
datanode1_1  |  at org.apache.ratis.thirdparty.io.netty.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:350)
datanode1_1  |  at org.apache.ratis.thirdparty.io.netty.handler.ssl.SslContextBuilder.forServer(SslContextBuilder.java:93)
datanode1_1  |  at org.apache.ratis.grpc.server.GrpcService.startBuildingNettyServer(GrpcService.java:227)
datanode1_1  |  at org.apache.ratis.grpc.server.GrpcService.<init>(GrpcService.java:191)
datanode1_1  |  at org.apache.ratis.grpc.server.GrpcService.<init>(GrpcService.java:122)
datanode1_1  |  at org.apache.ratis.grpc.server.GrpcService.<init>(GrpcService.java:53)
datanode1_1  |  at org.apache.ratis.grpc.server.GrpcService$Builder.build(GrpcService.java:74)
datanode1_1  |  at org.apache.ratis.grpc.GrpcFactory.newRaftServerRpc(GrpcFactory.java:121)
datanode1_1  |  at org.apache.ratis.grpc.GrpcFactory.newRaftServerRpc(GrpcFactory.java:39)
datanode1_1  |  at org.apache.ratis.server.impl.RaftServerProxy.<init>(RaftServerProxy.java:203)
datanode1_1  |  at org.apache.ratis.server.impl.ServerImplUtils.lambda$newRaftServer$0(ServerImplUtils.java:63)
datanode1_1  |  at org.apache.ratis.util.JavaUtils.attempt(JavaUtils.java:166)
datanode1_1  |  at org.apache.ratis.util.JavaUtils.attemptRepeatedly(JavaUtils.java:152)
datanode1_1  |  at org.apache.ratis.server.impl.ServerImplUtils.newRaftServer(ServerImplUtils.java:62)
datanode1_1  |  at org.apache.ratis.server.impl.ServerImplUtils.newRaftServer(ServerImplUtils.java:50)
datanode1_1  |  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
datanode1_1  |  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
datanode1_1  |  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
datanode1_1  |  at java.base/java.lang.reflect.Method.invoke(Method.java:566)
datanode1_1  |  at org.apache.ratis.server.RaftServer$Builder.newRaftServer(RaftServer.java:178)
datanode1_1  |  at org.apache.ratis.server.RaftServer$Builder.build(RaftServer.java:195)
datanode1_1  |  at org.apache.hadoop.ozone.container.common.transport.server.ratis.XceiverServerRatis.<init>(XceiverServerRatis.java:178)
datanode1_1  |  at org.apache.hadoop.ozone.container.common.transport.server.ratis.XceiverServerRatis.newXceiverServerRatis(XceiverServerRatis.java:450)
```

**Comment by adoroszlai** [comment]
> Just have reverted some changes in GrpcTlsConfig. Could you take a look?

Thanks @szetszwo, it works now.

File: ratis-common/src/main/java/org/apache/ratis/security/TlsConf.java [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    public final boolean isFileBased() {
+      return getFile() != null;
+    }
+  }
+
+  /** Configuration for {@link X509Certificate}s. */
+  public static class CertificatesConf extends FileBasedValue<Iterable<X509Certificate>> {
+    public CertificatesConf(Iterable<X509Certificate> certificates) {
+      super(certificates);
```
Comment (adoroszlai) [inline ratis-common/src/main/java/org/apache/ratis/security/TlsConf.java:68]: I think a check that `certificates` has at least one item would be useful.

File: ratis-common/src/main/java/org/apache/ratis/security/TlsConf.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      return privateKey;
+    }
+
+    /** @return the certificates for the private key. */
+    public CertificatesConf getKeyCertificates() {
+      return keyCertificates;
+    }
+  }
+
+  static final AtomicInteger COUNT = new AtomicInteger();
```
Comment (adoroszlai) [inline ratis-common/src/main/java/org/apache/ratis/security/TlsConf.java:126]: Nit: `COUNT` can be private.

**Review by adoroszlai — COMMENTED** [review]
Thanks @szetszwo for the patch, LGTM.  Two minor possible improvements noted.

**Comment by szetszwo** [comment]
@adoroszlai , thanks a lot for reviewing this.  Just have pushed a new commit for addressing your comments.

**Review by adoroszlai — APPROVED** [review]
Thanks @szetszwo for updating the patch, LGTM.

**Pull request merged**